File: NewRidgeFinancial2/hal_period_requirements.py
"""Widget period requirements vs loaded import cache — for HAL period planning."""

from __future__ import annotations

import json
from typing import Any

from import_cache_ttl import load_manifest, relevant_period_labels
from import_loader import load_import_bundle, quickbooks_import_dir, softdent_import_dir


WIDGET_PERIOD_RULES: dict[str, dict[str, Any]] = {
    "financialProductionTrend": {
        "label": "Production Trend & YTD",
        "needs": "current + prior calendar month in SoftDent dashboard",
        "datasets": ["softdent.dashboard"],
    },
    "financialPayerMix": {
        "label": "Payer Mix & Collections",
        "needs": "current + prior month SoftDent dashboard with collections/payer fields",
        "datasets": ["softdent.dashboard"],
    },
    "practiceFinancialOverview": {
        "label": "Practice Financial Overview",
        "needs": "current month QuickBooks revenue + SoftDent dashboard",
        "datasets": ["quickbooks.revenue", "softdent.dashboard"],
    },
    "quickbooksPlDetail": {
        "label": "QuickBooks P&L Detail",
        "needs": "current month QuickBooks revenue and expenses",
        "datasets": ["quickbooks.revenue", "quickbooks.expenses"],
    },
    "periodCloseAndPosting": {
        "label": "Period Close & Posting",
        "needs": "active document queue period (from document dates, not import CSV)",
        "datasets": ["local.documents"],
    },
    "claimsPipeline": {
        "label": "Claims Pipeline",
        "needs": "claim service dates (not monthly period labels)",
        "datasets": ["softdent.claims"],
    },
}


def _periods_from_rows(rows: list[dict[str, Any]] | None, field_names: tuple[str, ...]) -> list[str]:
    found: set[str] = set()
    for row in rows or []:
        if not isinstance(row, dict):
            continue
        for name in field_names:
            raw = row.get(name)
            if raw:
                found.add(str(raw).strip()[:7])
                break
    return sorted(found)


def analyze_widget_periods(bundle: dict[str, Any] | None = None) -> dict[str, Any]:
    bundle = bundle or load_import_bundle(sync=False, deep=False)
    required = relevant_period_labels()
    sd = (bundle.get("softdent") or {}) if isinstance(bundle, dict) else {}
    qb = (bundle.get("quickbooks") or {}) if isinstance(bundle, dict) else {}
    loaded = {
        "softdent.dashboard": _periods_from_rows((sd.get("dashboard") or {}).get("rows"), ("period", "Period")),
        "quickbooks.revenue": _periods_from_rows((qb.get("revenue") or {}).get("rows"), ("Period", "period")),
        "quickbooks.expenses": _periods_from_rows((qb.get("expenses") or {}).get("rows"), ("Period", "period")),
    }
    manifest = load_manifest() or {}
    widget_status: list[dict[str, Any]] = []
    for key, rule in WIDGET_PERIOD_RULES.items():
        dataset_keys = rule.get("datasets") or []
        if "local.documents" in dataset_keys:
            ok = True
            missing = []
        elif "softdent.claims" in dataset_keys:
            ok = bool(((sd.get("claims") or {}).get("rows") or []))
            missing = [] if ok else ["claims export rows"]
        else:
            needed_periods = required if any(k.startswith("softdent.") or k.startswith("quickbooks.") for k in dataset_keys) else required
            missing = []
            for ds in dataset_keys:
                have = set(loaded.get(ds) or [])
                for period in needed_periods:
                    if period not in have:
                        missing.append(f"{ds} missing {period}")
            ok = not missing
        widget_status.append(
            {
                "widgetKey": key,
                "label": rule["label"],
                "requirement": rule["needs"],
                "ok": ok,
                "missing": missing,
                "loadedPeriods": {ds: loaded.get(ds, []) for ds in dataset_keys if ds in loaded},
            }
        )
    return {
        "requiredPeriods": required,
        "loadedPeriods": loaded,
        "manifestPeriods": manifest.get("periods"),
        "widgets": widget_status,
        "policy": "Import cache keeps current + prior calendar month only (relevant-periods-only).",
        "softdentDir": str(softdent_import_dir()),
        "quickbooksDir": str(quickbooks_import_dir()),
    }


if __name__ == "__main__":
    print(json.dumps(analyze_widget_periods(), indent=2))

File: NewRidgeFinancial2/site/hal-period-requirements.js
/* global globalThis, window */
(function initHalPeriodRequirements(root) {
  "use strict";

  function relevantPeriodLabels(reference) {
    const now = reference || new Date();
    const current = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
    let y = now.getFullYear();
    let m = now.getMonth();
    if (m === 0) {
      m = 12;
      y -= 1;
    }
    const prior = `${y}-${String(m).padStart(2, "0")}`;
    return [current, prior];
  }

  function periodsFromRows(rows, fieldNames) {
    const found = new Set();
    (rows || []).forEach((row) => {
      if (!row || typeof row !== "object") return;
      fieldNames.forEach((name) => {
        const raw = row[name];
        if (raw) found.add(String(raw).trim().slice(0, 7));
      });
    });
    return Array.from(found).sort();
  }

  const WIDGET_PERIOD_RULES = {
    financialProductionTrend: {
      label: "Production Trend & YTD",
      needs: "current + prior calendar month in SoftDent dashboard",
      datasets: ["softdent.dashboard"],
    },
    financialPayerMix: {
      label: "Payer Mix & Collections",
      needs: "current + prior month SoftDent dashboard with collections/payer fields",
      datasets: ["softdent.dashboard"],
    },
    practiceFinancialOverview: {
      label: "Practice Financial Overview",
      needs: "current month QuickBooks revenue + SoftDent dashboard",
      datasets: ["quickbooks.revenue", "softdent.dashboard"],
    },
    quickbooksPlDetail: {
      label: "QuickBooks P&L Detail",
      needs: "current month QuickBooks revenue and expenses",
      datasets: ["quickbooks.revenue", "quickbooks.expenses"],
    },
    periodCloseAndPosting: {
      label: "Period Close & Posting",
      needs: "active document queue period (from document dates, not import CSV)",
      datasets: ["local.documents"],
    },
    claimsPipeline: {
      label: "Claims Pipeline",
      needs: "claim service dates (not monthly period labels)",
      datasets: ["softdent.claims"],
    },
  };

  function analyzeWidgetPeriods(snapshot) {
    const bundle = (snapshot && snapshot.importBundle) || {};
    const required = relevantPeriodLabels();
    const sd = bundle.softdent || {};
    const qb = bundle.quickbooks || {};
    const loaded = {
      "softdent.dashboard": periodsFromRows((sd.dashboard && sd.dashboard.rows) || [], ["period", "Period"]),
      "quickbooks.revenue": periodsFromRows((qb.revenue && qb.revenue.rows) || [], ["Period", "period"]),
      "quickbooks.expenses": periodsFromRows((qb.expenses && qb.expenses.rows) || [], ["Period", "period"]),
    };
    const widgets = Object.entries(WIDGET_PERIOD_RULES).map(([widgetKey, rule]) => {
      const datasetKeys = rule.datasets || [];
      let ok = true;
      let missing = [];
      if (datasetKeys.includes("local.documents")) {
        ok = Boolean(snapshot && snapshot.documents && snapshot.documents.queueCount);
      } else if (datasetKeys.includes("softdent.claims")) {
        ok = Boolean((sd.claims && sd.claims.rows && sd.claims.rows.length) || (snapshot && snapshot.claims && snapshot.claims.total));
        if (!ok) missing = ["claims export rows"];
      } else {
        datasetKeys.forEach((ds) => {
          const have = new Set(loaded[ds] || []);
          required.forEach((period) => {
            if (!have.has(period)) missing.push(`${ds} missing ${period}`);
          });
        });
        ok = missing.length === 0;
      }
      return {
        widgetKey,
        label: rule.label,
        requirement: rule.needs,
        ok,
        missing,
        loadedPeriods: Object.fromEntries(datasetKeys.filter((ds) => loaded[ds]).map((ds) => [ds, loaded[ds]])),
      };
    });
    return {
      requiredPeriods: required,
      loadedPeriods: loaded,
      widgets,
      policy: "Import cache keeps current + prior calendar month only (relevant-periods-only).",
    };
  }

  function formatWidgetPeriodRequirements(snapshot) {
    const analysis = analyzeWidgetPeriods(snapshot || {});
    const lines = [
      "Widget period requirements (local import cache only):",
      "",
      `Policy: ${analysis.policy}`,
      `Required calendar months: ${analysis.requiredPeriods.join(", ")}`,
      "",
      "Loaded periods by dataset:",
      `- SoftDent dashboard: ${(analysis.loadedPeriods["softdent.dashboard"] || []).join(", ") || "none"}`,
      `- QuickBooks revenue: ${(analysis.loadedPeriods["quickbooks.revenue"] || []).join(", ") || "none"}`,
      `- QuickBooks expenses: ${(analysis.loadedPeriods["quickbooks.expenses"] || []).join(", ") || "none"}`,
      "",
      "Widget coverage:",
    ];
    analysis.widgets.forEach((w) => {
      lines.push(`- [${w.ok ? "OK" : "GAP"}] ${w.label}: ${w.requirement}`);
      if (w.missing && w.missing.length) lines.push(`  Missing: ${w.missing.join("; ")}`);
    });
    lines.push("", "HAL uses mental time travel here: compare required months to loaded exports before recommending trend or close work.");
    return lines.join("\n");
  }

  const api = { relevantPeriodLabels, analyzeWidgetPeriods, formatWidgetPeriodRequirements };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  root.HalPeriodRequirements = api;
})(typeof globalThis !== "undefined" ? globalThis : typeof window !== "undefined" ? window : {});
